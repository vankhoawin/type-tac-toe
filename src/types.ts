import * as Enums from './enums';

export type Point = {
    row: number,
    col: number,
};
export type CurrentStatus = Enums.Status.Finished | Enums.Status.InProgress;
export type GameStateSquare = Enums.Square.Empty | Enums.Square.X | Enums.Square.O;
export type GameStateRow = [GameStateSquare, GameStateSquare, GameStateSquare];
export type GameStateGrid = [GameStateRow, GameStateRow, GameStateRow];
export type GameStateIdSelectors = {
    [key: string]: HTMLElement,
};
export type GameStateClassSelectors = {
    squares?: HTMLCollectionOf<Element>,
};
export type GameStateMeta = {
    status: CurrentStatus,
    turn: Enums.Turn,
    score: {
        player1: number,
        player2: number,
    },
};
export type GameEventHandlers = {
    startNewGame: (e: Event) => void,
    resetScore: (e: Event) => void,
    clickSquare: (e: Event) => void,
}
export type View = {
    selectors: GameStateIdSelectors,
    events: GameEventHandlers,
}
