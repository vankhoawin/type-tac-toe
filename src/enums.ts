export enum Turn {
    Player1 = 1,
    Player2 = 2,
}

export enum Status {
    InProgress = 0,
    Victory = 1,
    Draw = 2,
}

export enum Square {
    Empty = 0,
    X = 1,
    O = 2,
}
