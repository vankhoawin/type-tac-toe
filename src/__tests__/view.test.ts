import * as E from '../enums';
import * as T from '../types';
import View from '../view';
import scaffold from './scaffold';

describe('View', () => {
    let meta: T.IGameStateMeta;
    let state: T.GameStateGrid;
    let view: any;
    let startNewGameMock: jest.Mock;
    let resetScoreMock: jest.Mock;
    let clickSquareMock: jest.Mock;
    let element: string;

    beforeEach(() => {
        meta = scaffold.meta;
        state = scaffold.state;
        startNewGameMock = jest.fn();
        resetScoreMock = jest.fn();
        clickSquareMock = jest.fn();
        view = new View({
            events: {
                clickSquare: clickSquareMock,
                resetScore: resetScoreMock,
                startNewGame: startNewGameMock,
            },
            selectors: {
                board: document.getElementById('.board')!,
                toolbar: document.getElementById('.toolbar')!,
            },
        });
    });

    ['X', 'O', 'empty'].forEach((squareType, index) => {
        it(`renders an ${squareType} square`, () => {
            element = view.renderSquare(state[0][index], 0, index);

            expect(element).toMatchSnapshot();
        });

    });

    it('renders a grid', () => {
        element = view.renderGrid(state);

        expect(element).toMatchSnapshot();
    });

    it('renders a score subheader', () => {
        const turn: E.Turn = E.Turn.Player1;

        element = view.renderSubheader({ status: E.Status.Draw, turn });
        expect(element).toMatchSnapshot();

        element = view.renderSubheader({ status: E.Status.InProgress, turn });
        expect(element).toMatchSnapshot();

        element = view.renderSubheader({ status: E.Status.Victory, turn });
        expect(element).toMatchSnapshot();
    });

    it('renders a score container', () => {
        element = view.renderScore(meta.turn, meta.score[meta.turn]);

        expect(element).toMatchSnapshot();
    });

    it('renders toolbar buttons', () => {
        element = view.renderToolbarButtons();

        expect(element).toMatchSnapshot();
    });

    it('renders a toolbar', () => {
        element = view.renderToolbar(meta);

        expect(element).toMatchSnapshot();
    });
});
