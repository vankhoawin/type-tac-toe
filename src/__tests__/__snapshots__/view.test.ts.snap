// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View renders a grid 1`] = `"<button data-row=0 data-col=0 class=\\"js-board-square board__square board__square--X\\">X</button><button data-row=0 data-col=1 class=\\"js-board-square board__square board__square--O\\">O</button><button data-row=0 data-col=2 class=\\"js-board-square board__square board__square--empty\\"></button><button data-row=1 data-col=0 class=\\"js-board-square board__square board__square--empty\\"></button><button data-row=1 data-col=1 class=\\"js-board-square board__square board__square--X\\">X</button><button data-row=1 data-col=2 class=\\"js-board-square board__square board__square--empty\\"></button><button data-row=2 data-col=0 class=\\"js-board-square board__square board__square--empty\\"></button><button data-row=2 data-col=1 class=\\"js-board-square board__square board__square--empty\\"></button><button data-row=2 data-col=2 class=\\"js-board-square board__square board__square--empty\\"></button>"`;

exports[`View renders a score container 1`] = `
"
            <div class=\\"toolbar__score\\">
                Player 1: 3
            </div>
        "
`;

exports[`View renders a score subheader 1`] = `"Draw!"`;

exports[`View renders a score subheader 2`] = `"Player 1's turn."`;

exports[`View renders a score subheader 3`] = `"Player 1 wins!"`;

exports[`View renders a toolbar 1`] = `
"
            <div class=\\"toolbar__score-container\\">
                
            <div class=\\"toolbar__score\\">
                Player 1: 3
            </div>
        
                
            <div class=\\"toolbar__score\\">
                Player 2: 2
            </div>
        
            </div>
        
            <h2 class=\\"toolbar__subheader\\">
                Player 1's turn.
            </h2>
        
            <div class=\\"toolbar__button-container\\">
                <button
                    id=\\"js-new-button\\"
                    class=\\"toolbar__button\\"
                    type=\\"button\\"
                >
                    New Game
                </button>
                <button
                    id=\\"js-reset-button\\"
                    class=\\"toolbar__button\\"
                    type=\\"button\\"
                >
                    Reset Score
                </button>
            </div>
        "
`;

exports[`View renders an O square 1`] = `"<button data-row=0 data-col=1 class=\\"js-board-square board__square board__square--O\\">O</button>"`;

exports[`View renders an X square 1`] = `"<button data-row=0 data-col=0 class=\\"js-board-square board__square board__square--X\\">X</button>"`;

exports[`View renders an empty square 1`] = `"<button data-row=0 data-col=2 class=\\"js-board-square board__square board__square--empty\\"></button>"`;

exports[`View renders toolbar buttons 1`] = `
"
            <div class=\\"toolbar__button-container\\">
                <button
                    id=\\"js-new-button\\"
                    class=\\"toolbar__button\\"
                    type=\\"button\\"
                >
                    New Game
                </button>
                <button
                    id=\\"js-reset-button\\"
                    class=\\"toolbar__button\\"
                    type=\\"button\\"
                >
                    Reset Score
                </button>
            </div>
        "
`;
